### DISCUSSION TITLE
Refetching queries after mutation à la Apollo

### DISCUSSION THREAD

**Discussion by regirock365** [urql-graphql/urql#1517] [Q&A] [answered]
After working with Apollo for a while, I'm trying to pick up urql and having trouble understanding the different approach to updating the cache after mutations. In Apollo, I found this trivial to do by just providing the queries to be refetched upon completion like so:

```
const [ deleteOrderSession ] = useDeleteOrderSessionMutation({
    variables: { id },
    refetchQueries: [
      { query: OrderSessionsDocument },
      { query: OrderHistoryDocument },
      { query: AllClassesDocument },
    ],
  });
```

I'm unsure of the best way (if possible) to do this with urql. I'm providing mutation updates for Graphcache, but the update functions only provide the cache, which doesn't seem to have a query or refetch method. The alternatives are cache invalidation or manual updates which can feel like a hack in some situations.

is there a reason there isn't a refetch query method for mutation updates? Maybe there's something I'm missing?

Sorry in advance if my question is confusing.

**Comment by JoviDeCroock** [comment]
We handle this differently, we don't really agree with defining queries to refetch, we like to approach this scenario from a caching perspective.

When we are using our default cache, [`the document based cache`](https://formidable.com/open-source/urql/docs/basics/document-caching/) we approach queries/mutations from a `__typename` perspective, this means that if your `deleteOrder` returns something of `__typename: 'Order'` all those queries will be automatically invalidated and refetched. Additionally if you are just returning a `Boolean` for this then you can specify `additionalTypenames` to help you out `context: React.useMemo(['Order], [])`.

When we are talking about a [`normalized cache`](https://formidable.com/open-source/urql/docs/graphcache/) we rely on the user to specify `updater` function for deletion/addition and we'll handle updates for you.

**Comment by kitten** [comment, marked as answer]
To add on to @JoviDeCroock's answer and to address this;

> is there a reason there isn't a refetch query method for mutation updates? Maybe there's something I'm missing?

Yes, so ultimately it all comes down to avoiding coupling. Apollo encourages you to update this or that query or refetch it in place, when you define / send your mutation. Over time that introduces coupling and our reasoning here is: why would we want to express this in terms of queries when we may have a large number or changing ones?

So we think it's unacceptable to connect the two. This is especially true if you look at our normalised cache. If you'd eventually use it it also encourages you to define updates in a central config. All because your UI shouldn't control the cache but the cache and its data should "control" the UI. Having both introduces a lot of complications, maintenance burden, and confusion later on down the line

**Reply by nandorojo** [reply]
Perhaps this was added later, but I believe this is possible by invalidating the cache, right? https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#invalidating-entities
